pkg/restarter/restarter_test.go

```diff
@@ -0,0 +1,303 @@
+		{
+			Name: "Container-0",
+			State: v1.ContainerState{
+				Waiting: &v1.ContainerStateWaiting{
+					Reason:  "CrashLoopBackOff",
+					Message: "Container is in CrashLoopBackOff.",
+				},
+			},
+		},
+	}
+	return p
+}
+
+func newPodHealthy(name string, labels map[string]string) *v1.Pod {
+	p := testutil.NewPod(name, "node-0")
+	p.Labels = labels
+	p.Namespace = metav1.NamespaceDefault
+	p.Status.ContainerStatuses = []v1.ContainerStatus{
+		{
+			Name: "Container-0",
+		},
+	}
+	return p
+}
+
+func makePodUnhealthy(p *v1.Pod) *v1.Pod {
+	p.Status.ContainerStatuses = []v1.ContainerStatus{
+		{
+			Name: "Container-0",
+			State: v1.ContainerState{
+				Waiting: &v1.ContainerStateWaiting{
+					Reason:  "CrashLoopBackOff",
+					Message: "Container is in CrashLoopBackOff.",
+				},
+			},
+		},
+	}
+	return p
+}
+
+func (f *fixture) newController(deps *ServiceDependants, stopCh chan struct{}) (*Controller, informers.SharedInformerFactory, error) {
+
+	labelSelector := labels.Set(deps.Labels).AsSelector()
+	informers := informers.NewSharedInformerFactoryWithOptions(
+		f.client,
+		controller.NoResyncPeriodFunc(),
+		informers.WithNamespace(deps.Namespace),
+		informers.WithTweakListOptions(func(options *metav1.ListOptions) {
+			options.LabelSelector = labelSelector.String()
+		}))
+
+	c := NewController(f.client, informers, deps, watchDuration, stopCh)
+	for _, d := range f.eLister {
+		informers.Apps().V1().Deployments().Informer().GetIndexer().Add(d)
+	}
+	for _, pod := range f.pLister {
+		informers.Core().V1().Pods().Informer().GetIndexer().Add(pod)
+	}
+	return c, informers, nil
+}
+
+func TestDeleteOnlyCrashloopBackoffPods(t *testing.T) {
+	f := newFixture(t)
+	deps, err := DecodeConfigFile([]byte(dep))
+	if err != nil {
+		t.Fatalf("error decoding file: %v", err)
+	}
+	deps.Namespace = metav1.NamespaceDefault
+	stopCh := make(chan struct{})
+
+	e := newEndpoint(deps.Service, deps.Namespace, deps.Labels)
+	pC := newPodInCrashloop("pod-0", map[string]string{
+		"garden.sapcloud.io/role": "controlplane",
+		"role":                    "NotEtcd",
+	})
+	pH := newPodHealthy("pod-1", map[string]string{
+		"garden.sapcloud.io/role": "controlplane",
+		"role":                    "NotEtcd",
+	})
+
+	f.eLister = append(f.eLister, e)
+	//f.pLister = append(f.pLister, p)
+	f.objects = append(f.objects, e, pC, pH)
+	f.client = fake.NewSimpleClientset(f.objects...)
+	watcher := watch.NewFake()
+	f.client.PrependWatchReactor("pods", test.DefaultWatchReactor(watcher, nil))
+	// simulate add/update/delete watch events
+
+	c, _, err := f.newController(deps, stopCh)
+	if err != nil {
+		t.Fatalf("error creating Deployment controller: %v", err)
+	}
+	go func() {
+		pl, err := f.client.CoreV1().Pods(metav1.NamespaceDefault).List(metav1.ListOptions{})
+		if err != nil {
+			t.Fatalf("error fetching pods: %v", err)
+		}
+		before := len(pl.Items)
+		t.Logf("Number of pods before the watchdog started: %v.", before)
+		watcher.Add(pC)
+		watcher.Add(pH)
```

Comment:
Why not use the already existing mechanism in MCM tests? Even dispatching initial objects has been [implemented](https://github.com/gardener/machine-controller-manager/blob/master/pkg/fakeclient/client.go#L297:19).